Module not being loaded occasionally.

Sometimes the module is not working even after you force stop DANA.
If anybody got  a workaround for this, let me know!

Yes, tested in DANA v2.48.1. 
All the hook method is not found in this version but only one exception is thrown

Logcat:
```
--------- beginning of system
2023-12-18 12:41:52.476  2334-2446  VerityUtils             system_server                        E  Failed to measure fs-verity, errno 1: /data/app/~~1xq296RS_Wd-zBKmk8KtHg==/aimardcr.co.id.bypass_dana-rno0FIUiz6hslcYtjYN2Tw==/base.apk
--------- beginning of main
2023-12-18 12:41:52.650  2334-2446  VerityUtils             system_server                        E  Failed to measure fs-verity, errno 1: /data/app/~~1xq296RS_Wd-zBKmk8KtHg==/aimardcr.co.id.bypass_dana-rno0FIUiz6hslcYtjYN2Tw==/base.apk
2023-12-18 12:42:34.824 30046-30046 LSPosed-Bridge          pid-30046                            E  L.ckm.HAOP.m.HauXRRh.p.XposedHelpers$ClassNotFoundError: java.lang.ClassNotFoundException: id.dana.utils.IntegrityChecker
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findClass(Unknown Source:12)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findAndHookMethod(SourceFile:2)
                                                                                                    	at aimardcr.co.id.bypass_dana.NotSoUnsafe.handleLoadPackage(NotSoUnsafe.java:16)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.IXposedHookLoadPackage$Wrapper.handleLoadPackage(Unknown Source:2)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XC_LoadPackage.call(Unknown Source:6)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XCallback.callAll(Unknown Source:26)
                                                                                                    	at g0.a(Unknown Source:281)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at J.callback(Unknown Source:250)
                                                                                                    	at LSPHooker_.createOrUpdateClassLoaderLocked(Unknown Source:11)
                                                                                                    	at android.app.LoadedApk.getClassLoader(Unknown Source:8)
                                                                                                    	at android.app.LoadedApk.getResources(Unknown Source:50)
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:22)
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:1)
                                                                                                    	at android.app.ActivityThread.handleBindApplication(Unknown Source:630)
                                                                                                    	at android.app.ActivityThread.access$1600(Unknown Source:0)
                                                                                                    	at android.app.ActivityThread$H.handleMessage(Unknown Source:774)
                                                                                                    	at android.os.Handler.dispatchMessage(Unknown Source:19)
                                                                                                    	at android.os.Looper.loopOnce(Unknown Source:176)
                                                                                                    	at android.os.Looper.loop(Unknown Source:76)
                                                                                                    	at android.app.ActivityThread.main(Unknown Source:155)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(Unknown Source:11)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(Unknown Source:309)
                                                                                                    Caused by: java.lang.ClassNotFoundException: id.dana.utils.IntegrityChecker
                                                                                                    	at java.lang.Class.classForName(Native Method)
                                                                                                    	at java.lang.Class.forName(Class.java:536)
                                                                                                    	at q.a(Unknown Source:16)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findClass(Unknown Source:4)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findAndHookMethod(SourceFile:2) 
                                                                                                    	at aimardcr.co.id.bypass_dana.NotSoUnsafe.handleLoadPackage(NotSoUnsafe.java:16) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.IXposedHookLoadPackage$Wrapper.handleLoadPackage(Unknown Source:2) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XC_LoadPackage.call(Unknown Source:6) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XCallback.callAll(Unknown Source:26) 
                                                                                                    	at g0.a(Unknown Source:281) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at J.callback(Unknown Source:250) 
                                                                                                    	at LSPHooker_.createOrUpdateClassLoaderLocked(Unknown Source:11) 
                                                                                                    	at android.app.LoadedApk.getClassLoader(Unknown Source:8) 
                                                                                                    	at android.app.LoadedApk.getResources(Unknown Source:50) 
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:22) 
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:1) 
                                                                                                    	at android.app.ActivityThread.handleBindApplication(Unknown Source:630) 
                                                                                                    	at android.app.ActivityThread.access$1600(Unknown Source:0) 
                                                                                                    	at android.app.ActivityThread$H.handleMessage(Unknown Source:774) 
                                                                                                    	at android.os.Handler.dispatchMessage(Unknown Source:19) 
                                                                                                    	at android.os.Looper.loopOnce(Unknown Source:176) 
                                                                                                    	at android.os.Looper.loop(Unknown Source:76) 
                                                                                                    	at android.app.ActivityThread.main(Unknown Source:155) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(Unknown Source:11) 
                                                                                                    	at com.android.internal.os.ZygoteInit.main(Unknown Source:309) 
                                                                                                    Caused by: java.lang.ClassNotFoundException: id.dana.utils.IntegrityChecker
                                                                                                    	at java.lang.Class.classForName(Native Method) 
                                                                                                    	at java.lang.Class.forName(Class.java:536) 
                                                                                                    	at q.a(Unknown Source:16) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findClass(Unknown Source:4) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findAndHookMethod(SourceFile:2) 
                                                                                                    	at aimardcr.co.id.bypass_dana.NotSoUnsafe.handleLoadPackage(NotSoUnsafe.java:16) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.IXposedHookLoadPackage$Wrapper.handleLoadPackage(Unknown Source:2) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XC_LoadPackage.call(Unknown Source:6) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XCallback.callAll(Unknown Source:26) 
                                                                                                    	at g0.a(Unknown Source:281) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at J.callback(Unknown Source:250) 
                                                                                                    	at LSPHooker_.createOrUpdateClassLoaderLocked(Unknown Source:11) 
                                                                                                    	at android.app.LoadedApk.getClassLoader(Unknown Source:8) 
                                                                                                    	at android.app.LoadedApk.getResources(Unknown Source:50) 
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:22) 
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:1) 
                                                                                                    	at android.app.ActivityThread.handleBindApplication(Unknown Source:630) 
                                                                                                    	at android.app.ActivityThread.access$1600(Unknown Source:0) 
                                                                                                    	at android.app.ActivityThread$H.handleMessage(Unknown Source:774) 
                                                                                                    	at android.os.Handler.dispatchMessage(Unknown Source:19) 
                                                                                                    	at android.os.Looper.loopOnce(Unknown Source:176) 
                                                                                                    	at android.os.Looper.loop(Unknown Source:76) 
                                                                                                    	at android.app.ActivityThread.main(Unknown Source:155) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(Unknown Source:11) 
                                                                                                    	at com.android.internal.os.ZygoteInit.main(Unknown Source:309) 
2023-12-18 12:42:36.428 30354-30354 LSPosed-Bridge          pid-30354                            E  L.ckm.HAOP.m.HauXRRh.p.XposedHelpers$ClassNotFoundError: java.lang.ClassNotFoundException: id.dana.utils.IntegrityChecker
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findClass(Unknown Source:12)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findAndHookMethod(SourceFile:2)
                                                                                                    	at aimardcr.co.id.bypass_dana.NotSoUnsafe.handleLoadPackage(NotSoUnsafe.java:16)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.IXposedHookLoadPackage$Wrapper.handleLoadPackage(Unknown Source:2)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XC_LoadPackage.call(Unknown Source:6)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XCallback.callAll(Unknown Source:26)
                                                                                                    	at g0.a(Unknown Source:281)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at J.callback(Unknown Source:250)
                                                                                                    	at LSPHooker_.createOrUpdateClassLoaderLocked(Unknown Source:11)
                                                                                                    	at android.app.LoadedApk.getClassLoader(Unknown Source:8)
                                                                                                    	at android.app.LoadedApk.getResources(Unknown Source:50)
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:22)
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:1)
                                                                                                    	at android.app.ActivityThread.handleBindApplication(Unknown Source:630)
                                                                                                    	at android.app.ActivityThread.access$1600(Unknown Source:0)
                                                                                                    	at android.app.ActivityThread$H.handleMessage(Unknown Source:774)
                                                                                                    	at android.os.Handler.dispatchMessage(Unknown Source:19)
                                                                                                    	at android.os.Looper.loopOnce(Unknown Source:176)
                                                                                                    	at android.os.Looper.loop(Unknown Source:76)
                                                                                                    	at android.app.ActivityThread.main(Unknown Source:155)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(Unknown Source:11)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(Unknown Source:309)
                                                                                                    Caused by: java.lang.ClassNotFoundException: id.dana.utils.IntegrityChecker
                                                                                                    	at java.lang.Class.classForName(Native Method)
                                                                                                    	at java.lang.Class.forName(Class.java:536)
                                                                                                    	at q.a(Unknown Source:16)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findClass(Unknown Source:4)
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findAndHookMethod(SourceFile:2) 
                                                                                                    	at aimardcr.co.id.bypass_dana.NotSoUnsafe.handleLoadPackage(NotSoUnsafe.java:16) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.IXposedHookLoadPackage$Wrapper.handleLoadPackage(Unknown Source:2) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XC_LoadPackage.call(Unknown Source:6) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XCallback.callAll(Unknown Source:26) 
                                                                                                    	at g0.a(Unknown Source:281) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at J.callback(Unknown Source:250) 
                                                                                                    	at LSPHooker_.createOrUpdateClassLoaderLocked(Unknown Source:11) 
                                                                                                    	at android.app.LoadedApk.getClassLoader(Unknown Source:8) 
                                                                                                    	at android.app.LoadedApk.getResources(Unknown Source:50) 
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:22) 
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:1) 
                                                                                                    	at android.app.ActivityThread.handleBindApplication(Unknown Source:630) 
                                                                                                    	at android.app.ActivityThread.access$1600(Unknown Source:0) 
                                                                                                    	at android.app.ActivityThread$H.handleMessage(Unknown Source:774) 
                                                                                                    	at android.os.Handler.dispatchMessage(Unknown Source:19) 
                                                                                                    	at android.os.Looper.loopOnce(Unknown Source:176) 
                                                                                                    	at android.os.Looper.loop(Unknown Source:76) 
                                                                                                    	at android.app.ActivityThread.main(Unknown Source:155) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(Unknown Source:11) 
                                                                                                    	at com.android.internal.os.ZygoteInit.main(Unknown Source:309) 
                                                                                                    Caused by: java.lang.ClassNotFoundException: id.dana.utils.IntegrityChecker
                                                                                                    	at java.lang.Class.classForName(Native Method) 
                                                                                                    	at java.lang.Class.forName(Class.java:536) 
                                                                                                    	at q.a(Unknown Source:16) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findClass(Unknown Source:4) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.XposedHelpers.findAndHookMethod(SourceFile:2) 
                                                                                                    	at aimardcr.co.id.bypass_dana.NotSoUnsafe.handleLoadPackage(NotSoUnsafe.java:16) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.IXposedHookLoadPackage$Wrapper.handleLoadPackage(Unknown Source:2) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XC_LoadPackage.call(Unknown Source:6) 
                                                                                                    	at L.ckm.HAOP.m.HauXRRh.p.callbacks.XCallback.callAll(Unknown Source:26) 
                                                                                                    	at g0.a(Unknown Source:281) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at J.callback(Unknown Source:250) 
                                                                                                    	at LSPHooker_.createOrUpdateClassLoaderLocked(Unknown Source:11) 
                                                                                                    	at android.app.LoadedApk.getClassLoader(Unknown Source:8) 
                                                                                                    	at android.app.LoadedApk.getResources(Unknown Source:50) 
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:22) 
                                                                                                    	at android.app.ContextImpl.createAppContext(Unknown Source:1) 
                                                                                                    	at android.app.ActivityThread.handleBindApplication(Unknown Source:630) 
                                                                                                    	at android.app.ActivityThread.access$1600(Unknown Source:0) 
                                                                                                    	at android.app.ActivityThread$H.handleMessage(Unknown Source:774) 
                                                                                                    	at android.os.Handler.dispatchMessage(Unknown Source:19) 
                                                                                                    	at android.os.Looper.loopOnce(Unknown Source:176) 
                                                                                                    	at android.os.Looper.loop(Unknown Source:76) 
                                                                                                    	at android.app.ActivityThread.main(Unknown Source:155) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(Unknown Source:11) 
                                                                                                    	at com.android.internal.os.ZygoteInit.main(Unknown Source:309) 
```

--- a/app/src/main/java/aimardcr/co/id/bypass_dana/NotSoUnsafe.java
+++ b/app/src/main/java/aimardcr/co/id/bypass_dana/NotSoUnsafe.java
@@ -1,51 +1,140 @@
 package aimardcr.co.id.bypass_dana;
 
+import android.app.Application;
+import android.content.Context;
+import android.widget.Toast;
+
+import androidx.annotation.NonNull;
+
 import de.robv.android.xposed.IXposedHookLoadPackage;
 import de.robv.android.xposed.XC_MethodHook;
-import de.robv.android.xposed.XposedBridge;
 import de.robv.android.xposed.XposedHelpers;
 import de.robv.android.xposed.callbacks.XC_LoadPackage;
 
 public class NotSoUnsafe implements IXposedHookLoadPackage {
+
+    // target app classloader & context
+    private ClassLoader pClassLoader;
+    private Context context;
+
     @Override
-    public void handleLoadPackage(final XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {
+    public void handleLoadPackage(@NonNull final XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {
         if (!lpparam.packageName.equals("id.dana")) {
             return;
         }
 
-        XposedHelpers.findAndHookMethod("id.dana.utils.IntegrityChecker", lpparam.classLoader, "ArraysUtil", String.class, new XC_MethodHook() {
-            @Override
-            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
-                param.setResult(null);
-            }
-        });
+        pClassLoader = lpparam.classLoader;
+        initAppContext();
+    }
+    public void initHook() {
+        integrityCheck();
+        emuCheck();
+        hookCheck();
+        rootCheck();
+        tampCheck();
+    }
 
-        XposedHelpers.findAndHookMethod("id.dana.DanaApplication", lpparam.classLoader, "emuCb", long.class, long.class, new XC_MethodHook() {
-            @Override
-            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
-                param.setResult(null);
-            }
-        });
+    public void integrityCheck() {
+        if (pClassLoader == null || context == null) {
+            throw new NullPointerException("pClassLoader/context is null");
+        }
 
-        XposedHelpers.findAndHookMethod("id.dana.DanaApplication", lpparam.classLoader, "hookCb", long.class, long.class, new XC_MethodHook() {
-            @Override
-            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
-                param.setResult(null);
-            }
-        });
+        try {
+            XposedHelpers.findAndHookMethod("id.dana.utils.IntegrityChecker", pClassLoader, "ArraysUtil", String.class, new XC_MethodHook() {
+                @Override
+                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
+                    param.setResult(null);
+                }
+            });
+        } catch (Throwable e) {
+            e.printStackTrace();
+            Toast.makeText(context, "Unable to bypass integrityCheck!", Toast.LENGTH_SHORT).show();
+        }
+    }
 
-        XposedHelpers.findAndHookMethod("id.dana.DanaApplication", lpparam.classLoader, "rootCb", long.class, long.class, new XC_MethodHook() {
-            @Override
-            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
-                param.setResult(null);
-            }
-        });
+    public void emuCheck() {
+        if (pClassLoader == null || context == null) {
+            throw new NullPointerException("pClassLoader/context is null");
+        }
 
-        XposedHelpers.findAndHookMethod("id.dana.DanaApplication", lpparam.classLoader, "tampCb", long.class, long.class, new XC_MethodHook() {
-            @Override
-            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
-                param.setResult(null);
-            }
-        });
+        try {
+            XposedHelpers.findAndHookMethod("id.dana.DanaApplication", pClassLoader, "emuCb", long.class, long.class, new XC_MethodHook() {
+                @Override
+                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
+                    param.setResult(null);
+                }
+            });
+        } catch (Throwable e) {
+            e.printStackTrace();
+            Toast.makeText(context, "Unable to bypass emuCheck!", Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    public void hookCheck() {
+        if (pClassLoader == null || context == null) {
+            throw new NullPointerException("pClassLoader/context is null");
+        }
+
+        try {
+            XposedHelpers.findAndHookMethod("id.dana.DanaApplication", pClassLoader, "hookCb", long.class, long.class, new XC_MethodHook() {
+                @Override
+                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
+                    param.setResult(null);
+                }
+            });
+        } catch (Throwable e) {
+            e.printStackTrace();
+            Toast.makeText(context, "Unable to bypass hookCheck!", Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    public void rootCheck() {
+        if (pClassLoader == null || context == null) {
+            throw new NullPointerException("pClassLoader/context is null");
+        }
+
+        try {
+            XposedHelpers.findAndHookMethod("id.dana.DanaApplication", pClassLoader, "rootCb", long.class, long.class, new XC_MethodHook() {
+                @Override
+                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
+                    param.setResult(null);
+                }
+            });
+        } catch (Throwable e) {
+            e.printStackTrace();
+            Toast.makeText(context, "Unable to bypass rootCheck!", Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    public void tampCheck() {
+        if (pClassLoader == null || context == null) {
+            throw new NullPointerException("pClassLoader/context is null");
+        }
+
+        try {
+            XposedHelpers.findAndHookMethod("id.dana.DanaApplication", pClassLoader, "tampCb", long.class, long.class, new XC_MethodHook() {
+                @Override
+                protected void beforeHookedMethod(XC_MethodHook.MethodHookParam param) throws Throwable {
+                    param.setResult(null);
+                }
+            });
+        } catch (Throwable e) {
+            e.printStackTrace();
+            Toast.makeText(context, "Unable to bypass tampCheck!", Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    public void initAppContext() {
+        try {
+            XposedHelpers.findAndHookMethod(Application.class.getName(), null, "attach", Context.class, new XC_MethodHook() {
+                @Override
+                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
+                    context = (Context) param.args[0];
+                    initHook();
+                }
+            });
+        } catch (Throwable e) {
+            e.printStackTrace();
+        }
     }
 }

